nova/virt/vmwareapi/vmops.py

```diff
@@ -3125,3 +3134,119 @@ def update_vmref_cache(self):
                 continue
 
             vm_util.vm_ref_cache_update(vm_uuid, props['obj'])
+
+    def set_compute_host(self, compute_host):
+        """Called by the driver on init_host() so we know the compute host"""
+        self._compute_host = compute_host
+
+    def sync_server_group(self, context, sg_uuid):
+        # we have to ignore instances currently in a volatitle state, where
+        # either VMware cannot support them being in a DRS rule or we expect
+        # them to go away during the syncing process, which could lead to
+        # errors.
+        IGNORED_INSTANCES_STATES = [
+            task_states.MIGRATING,
+            task_states.DELETING,
+            task_states.SHELVING,
+            task_states.REBUILDING,
+            task_states.REBUILD_BLOCK_DEVICE_MAPPING,
+        ]
```

Comment:
@fwiesel From a quick glance over the task states, these are the ones I'd like to exclude from the DRS rule.

Reply:
I think `IGNORED_INSTANCES_STATES` is not the right term, it is `REMOVE_INSTANCE_GROUP_STATES` or something like that.
I would leave it at `MIGRATING` and `DELETING`. `REBUILDING` creates a new instance, for which we potentially want to keep the instance group.

Reply:
Wasn't sure about the naming. `REMOVE_INSTANCE_GROUP_STATES` sounds like we remove instance groups in certain states. Technically, we actually ignore those members, so maybe `IGNORE_MEMBERS_IN_STATES` or something?

The point with `REBUILDING` is, that it will remove the current instance from the cluster and thus our reconfigure might fail, because the `moref` we're trying to sync is not available in the cluster anymore. Basically depends on which part of `REBUILDING` we're looking at, if we want to keep the member or not - and for that we have `REBUILD_SPAWNING`, which we don't ignore and where we can add the member again.

I thought, setting `expected_attrs` to something would actually make the default joins disappear, as `columns_to_join` was generated out of those. `task_state` should be in the `instances` table, so I didn't see a reason to add it to `expected_attrs`, because that parameter is only to determine if tables should be joined.

I'll check again though.

Reply:
> Wasn't sure about the naming. REMOVE_INSTANCE_GROUP_STATES sounds like we remove instance groups in certain states. Technically, we actually ignore those members, so maybe IGNORE_MEMBERS_IN_STATES or something?

Ah, it depends from which perspective you consider it. I am considering it from the perspective of the vcenter rules.
From that perspective, they are not ignored, but removed. I assume, you are talking then from the perspective of nova instance group.
Maybe then we call `EXCLUDE_FROM_DRS_RULES` or something like that?
I think, it is important to be clear, that actually the sync does unapply the Nova DRS rules for that instance

For the migration, that is what I want.

> The point with REBUILDING is [...]

Fine by me then.

> I thought, setting `expected_attrs` to something [...]

I think, you are right. I was thinking it was a way to specify which columns to select, in order to avoid to load user-data, keys, etc... That doesn't seem to be integrated.

Reply:
Checked it again: We do not control `columns_to_join` directly, but can only pass `expected_attr`, which is [used to generate](https://github.com/sapcc/nova/blob/d388e392d7edee6c140969e962af2347b34bf568/nova/objects/instance.py#L1327) `columns_to_join` [here](https://github.com/sapcc/nova/blob/d388e392d7edee6c140969e962af2347b34bf568/nova/objects/instance.py#L1327). That's why we pass `expected_attr=[]`, so `columns_to_join` becomes `[]`.

Reply:
I named it `STATES_EXCLUDING_MEMBERS_FROM_DRS_RULES`, which should make both of us happy, I think. It also contains a hint on what's contained in the list.